### PR TITLE
Fon und kiwi

### PR DESCRIPTION
**Pull request by zaphod-bx**
kiwi!

### CODE REVIEW

File: Documentation/ProjectFile/MangaProject/tree_dynamics/belowground_competition/FON/s_salinity.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
+
+salinity in ppt.
```
Comment (jbathmann) [inline Documentation/ProjectFile/MangaProject/tree_dynamics/belowground_competition/FON/s_salinity.md]: For example: do I need to input 35 ppt or 0.035?
Please state here...

File: PopulationLib/Species/Avicennia.py [hunk 2]
```diff
@@ -24,6 +24,14 @@ def createTree():
     growth_factor = 0.0015
     h_sigmo_slope = 0.5
     sigmo_slope = 0.15
+    # Kiwi
```
Comment (jbathmann) [inline PopulationLib/Species/Avicennia.py]: maybe one blsnk line above and one below the kiwi block?

File: PopulationLib/Species/Avicennia.py [hunk 3]
```diff
@@ -42,4 +50,13 @@ def createTree():
     parameter["growth_factor"] = growth_factor
     parameter["h_sigmo_slope"] = h_sigmo_slope
     parameter["sigmo_slope"] = sigmo_slope
+    parameter["salt_effect_d"] = salt_effect_d
```
Comment (jbathmann) [inline PopulationLib/Species/Avicennia.py]: indicate, that the kiwi_block is started

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+@date: 2018-Today
+@author: jasper.bathmann@ufz.de
+"""
+from TreeModelLib.GrowthAndDeathDynamics import GrowthAndDeathDynamics
+
+
+class SimpleKiwi(GrowthAndDeathDynamics):
+    ## SimpleKiwi for death and growth dynamics. For details see
+    #  https://doi.org/10.1016/S0304-3800(00)00298-2 \n
+    #  @param Tags to define SimpleBettina: type
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: >???

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    ## This functions is the main routine for reading the tree geometry and
+    #  parameters, scheduling the computations and updating the tree geometry.\n
+    #  @param tree - object of type tree\n
+    #  @param aboveground_resources - fraction of maximum light interception (shading effect)\n
+    #  @param belowground_resources - fract of max water upt (compet and/or salinity > 0)
+    def progressTree(self, tree, aboveground_resources, belowground_resources):
+        geometry = tree.getGeometry()
+        parameter = tree.getParameter()
+        tree.setGeometry(geometry)
+        tree.setSurvival(1)
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: Group this with lines 44 and 45

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        height = (137 + parameter["b2"] * dbh - parameter["b3"] * dbh**2)
+        growth = ( parameter["max_growth"] * dbh * 
+           (1 - dbh * height / parameter["max_dbh"] / parameter["max_height"]) /
+           (274 + 3 * parameter["b2"] * dbh - 4 * parameter["b3"] * dbh**2)
+            * belowground_resources) 
+        dbh = dbh + growth * (self.t_end - self.t_ini)/(3600*24*365)
+        if growth < parameter["mortality_constant"]:
+            tree.setSurvival(0)
+        print(growth)
+        print(dbh)
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: please drop

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        parameter = tree.getParameter()
+        tree.setGeometry(geometry)
+        tree.setSurvival(1)
+        dbh = geometry["r_stem"] * 200
+        height = (137 + parameter["b2"] * dbh - parameter["b3"] * dbh**2)
+        growth = ( parameter["max_growth"] * dbh * 
+           (1 - dbh * height / parameter["max_dbh"] / parameter["max_height"]) /
+           (274 + 3 * parameter["b2"] * dbh - 4 * parameter["b3"] * dbh**2)
+            * belowground_resources) 
+        dbh = dbh + growth * (self.t_end - self.t_ini)/(3600*24*365)
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: I trust you ;-)

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self, args):
+        case = args.find("type").text
+        print("Initiate belowground competition of type " + case + ".")
+
+    ## This functions prepares the growth and death concept.
+    #  In the SimpleKiwi concept, the timestepping is updated. 
+    #  @param t_ini - initial time for next timestep \n
+    #  @param t_end - end time for next
+    def prepareNextTimeStep(self, t_ini, t_end):
+        self.t_ini = t_ini
```
Comment (jbathmann) [inline TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: please change all internal variables to private members:
self.t_ini -> self._t_ini

File: TreeModelLib/BelowgroundCompetition/FON/FON.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+"""
+import numpy as np
+from TreeModelLib.BelowgroundCompetition import BelowgroundCompetition
+
+
+class FON(BelowgroundCompetition):
+    ## FON case for belowground competition concept. For details see
+    #  (https://doi.org/10.1016/S0304-3800(00)00298-2). FON returns a list 
+    #  of multipliers for each tree for salinity and competition.\n
+    #  @param: Tags to define FON: type\n
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/FON/FON.py]: what about the other parameters?

File: TreeModelLib/BelowgroundCompetition/FON/FON.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/FON/FON.py]: please use private members:

self.aa -> self._aa

File: TreeModelLib/BelowgroundCompetition/FON/FON.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        self.r_stem = []
+        self.t_ini = t_ini
+        self.t_end = t_end
+        self.fon_height = np.zeros_like(self.my_grid[0])
+
+    ## Before being able to calculate the resources, all tree entities need
+    #  to be added with their relevant allometric measures for the next timestep.
+    #  @param: position, geometry, parameter
+    def addTree(self, x, y, geometry, parameter):
+        self.trees.append(1)
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/FON/FON.py]: do you need to use the self.trees list? drop?

File: TreeModelLib/BelowgroundCompetition/FON/FON.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    def __init__(self, args):
+        case = args.find("type").text
+        print("Initiate belowground competition of type " + case + ".")
+        self.makeGrid(args)
+
+    ## This function returns a list of the growth reduction factors of all trees.
+    #  calculated in the subsequent timestep.\n  
+    #  @return: np.array with $N_tree$ scalars
+    def calculateBelowgroundResources(self):
+        for i in range(len(self.xe)):
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/FON/FON.py]: maybe, for readability:

`n_trees = len(self.xe)`

and than using this for the for loops

File: Documentation/ProjectFile/MangaProject/tree_dynamics/belowground_competition/FON/s_salinity.md (hunk 1 above)
Comment (zaphod-bx) [reply Documentation/ProjectFile/MangaProject/tree_dynamics/belowground_competition/FON/s_salinity.md]: in ppt means 35

File: PopulationLib/Species/Avicennia.py (hunk 2 above)
Comment (zaphod-bx) [reply PopulationLib/Species/Avicennia.py]: okay

File: PopulationLib/Species/Avicennia.py (hunk 3 above)
Comment (zaphod-bx) [reply PopulationLib/Species/Avicennia.py]: okay

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py (hunk 4 above)
Comment (zaphod-bx) [reply TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: okay

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py (hunk 5 above)
Comment (zaphod-bx) [reply TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: okay

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py (hunk 6 above)
Comment (zaphod-bx) [reply TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: done

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py (hunk 7 above)
Comment (zaphod-bx) [reply TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: fine

File: TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py (hunk 8 above)
Comment (zaphod-bx) [reply TreeModelLib/GrowthAndDeathDynamics/SimpleKiwi/SimpleKiwi.py]: krass, geht!

File: TreeModelLib/BelowgroundCompetition/FON/FON.py (hunk 9 above)
Comment (zaphod-bx) [reply TreeModelLib/BelowgroundCompetition/FON/FON.py]: juti

File: TreeModelLib/BelowgroundCompetition/FON/FON.py (hunk 10 above)
Comment (zaphod-bx) [reply TreeModelLib/BelowgroundCompetition/FON/FON.py]: yupp

File: TreeModelLib/BelowgroundCompetition/FON/FON.py (hunk 11 above)
Comment (zaphod-bx) [reply TreeModelLib/BelowgroundCompetition/FON/FON.py]: jut

File: TreeModelLib/BelowgroundCompetition/FON/FON.py (hunk 12 above)
Comment (zaphod-bx) [reply TreeModelLib/BelowgroundCompetition/FON/FON.py]: is ferc

**Pull request merged**